drivers/gpio/gpio_pca95xx.c

```diff
@@ -448,12 +484,149 @@ static int gpio_pca95xx_port_toggle_bits(const struct device *dev,
 	return ret;
 }
 
+static void gpio_pca95xx_interrupt_worker(struct k_work *work)
+{
+	struct gpio_pca95xx_drv_data * const drv_data = CONTAINER_OF(
+		work, struct gpio_pca95xx_drv_data, interrupt_worker);
+	uint16_t input_status, input_cache;
+	uint32_t enabled_int = 0;
+	int ret;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	input_cache = drv_data->reg_cache.input;
+
+	ret = update_input_regs(drv_data->instance, &input_status);
+	if (ret == 0) {
+		/* Note: PCA Interrupt status is cleared by reading inputs */
+
+		enabled_int = ((input_cache ^ input_status) &
+			       drv_data->pin_callback_enables);
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	gpio_fire_callbacks(&drv_data->callbacks, drv_data->instance,
+			    enabled_int);
+}
+
+static void gpio_pca95xx_interrupt_callback(const struct device *dev,
+					    struct gpio_callback *cb,
+					    gpio_port_pins_t pins)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		CONTAINER_OF(cb, struct gpio_pca95xx_drv_data, gpio_callback);
+
+	/* Cannot read PCA95xx registers from ISR context, queue worker */
+	k_work_submit(&drv_data->interrupt_worker);
+}
+
 static int gpio_pca95xx_pin_interrupt_configure(const struct device *dev,
 						  gpio_pin_t pin,
 						  enum gpio_int_mode mode,
 						  enum gpio_int_trig trig)
 {
-	return -ENOTSUP;
+	const struct gpio_pca95xx_config * const config = dev->config;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->data;
+	const struct device *int_gpio_dev;
+	int ret = 0;
+	uint16_t reg;
+
+	/* Check if GPIO port supports interrupts */
+	if ((config->capabilities & PCA_HAS_INTERRUPT) == 0U) {
+		return -ENOTSUP;
+	}
+
+	/* Check for an invalid pin number */
+	if (pin > config->common.port_pin_mask) {
+		return -EINVAL;
+	}
+
+	/* PCA95xx only supports edge triggered interrupts */
+	if ((mode != GPIO_INT_MODE_DISABLED) &&
+	    ((mode & GPIO_INT_EDGE) == 0)) {
+		LOG_ERR("PCA95XX[0x%X]: only edge trig. interrupts supported",
+			config->i2c_slave_addr);
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Check if we should skip interrupt initizalization */
+	if ((mode == GPIO_INT_MODE_DISABLED) ||
+	    (drv_data->interrupt_initialized)) {
+		goto done;
+	}
+
+	/* Prepare interrupt worker */
+	k_work_init(&drv_data->interrupt_worker,
+		    gpio_pca95xx_interrupt_worker);
+
+	/* Enable GPIO callback for interrupt pin */
+	int_gpio_dev = device_get_binding(config->int_gpio_port);
+	if (int_gpio_dev == NULL) {
+		LOG_ERR("PCA95XX[0x%X]: error getting interrupt GPIO device"
+			" (%s)", config->i2c_slave_addr,
+			config->int_gpio_port);
+		ret = -ENODEV;
+		goto err;
+	}
+
+	ret = gpio_pin_configure(int_gpio_dev, config->int_gpio_pin,
```

Comment:
This doesn't seem ideal.  You've got the logic to configure the SoC GPIO input on which the PCA95xx will signal interrupts within the code to configure interrupts on a PCA95xx pin.

Configuration of that pin and the callback should be done when the driver is initialized.  The pin itself should be configured to interrupt mode if and only if there is at least one PCA95xx pin that has been configured for interrupt.

Putting the logic here, and leaving the interrupt enabled when there are no clients, is not ideal---especially if it's true that the PCA95xx will signal state changes on every input change.

Reply:
configuration moved to init(), INT pin interrupt is enabled / disabled as needed